### PR TITLE
tedge init command #1764

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

- [x] Introduce `tedge init` command to initialise directories/files requiring elevated privileges deprecating the old `tedge --init` option. The old `--init` option was replaced with the `init` subcommand to support additional `--user` and `--group` options seamlessly
- [x] Make all tedge daemons create the necessary directories/files on their startup
- [x] Deprecate the `--init` command from all tedge daemons with a warning. Not removing them yet to avoid breaking backward compatibility

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/main.rs [hunk 1]
```diff
@@ -4,6 +4,7 @@ use clap::Parser;
 use tedge_config::system_services::get_log_level;
 use tedge_config::system_services::set_log_level;
 use tedge_config::DEFAULT_TEDGE_CONFIG_PATH;
+use tracing::log::warn;
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/main.rs:7]: Is using trace a good idea jut to log a `warning`?

**Review by PradeepKiruvale — COMMENTED** [review]
- I did not get why we need to expose creating `user` and `group` with `init` command, because there are only two users/group used by the `thin-edge`, one is `tedge` user and the other is `mosquitto`, which is already created by mosquito itself.

- `init/clear_session` is not relevant anymore? If so, these need to be removed these implementations from the library as well right?

**Comment by albinsuresh** [comment]
> * I did not get why we need to expose creating `user` and `group` with `init` command, because there are only two users/group used by the `thin-edge`, one is `tedge` user and the other is `mosquitto`, which is already created by mosquito itself.

The `--user` and `--group` options are not there to create those users, but to define which user/group must own the files/directories created. The user should have been created separately(happens in the `preinst` script of the `tedge` debian package). Provided these options so that we are not tied to just `tedge` user even at the code level, but the user can define any user for the same. For security, we just want all tedge services to run as a specific user owing things as that user. It doesn't matter if that user is called `tedge` or whatever. We just use `tedge` as the default.

> * `init/clear_session` is not relevant anymore? If so, these need to be removed these implementations from the library as well right?

You're right. I thought I had cleared all the `init` APIs. Will check again and will remove `clear_session` as well.

File: crates/core/tedge_agent/src/main.rs (hunk 1 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/main.rs:7]: I just used tracing as `tedge-agent` is already using the `tracing` crate for logging and I didn't want to add `log` as well as an additonal dependency. May be we should migrate all the crates to use just `log` as we had agreed a while ago, but in a separate PR.

**Comment by PradeepKiruvale** [comment]
> > * I did not get why we need to expose creating `user` and `group` with `init` command, because there are only two users/group used by the `thin-edge`, one is `tedge` user and the other is `mosquitto`, which is already created by mosquito itself.
> 
> The `--user` and `--group` options are not there to create those users, but to define which user/group must own the files/directories created. The user should have been created separately(happens in the `preinst` script of the `tedge` debian package). Provided these options so that we are not tied to just `tedge` user even at the code level, but the user can define any user for the same. For security, we just want all tedge services to run as a specific user owing things as that user. It doesn't matter if that user is called `tedge` or whatever. We just use `tedge` as the default.

Got it, but when all the services are on the same `node`, If the user gives different `group` names won't this affect creating the specific directories/files?
> 
> > * `init/clear_session` is not relevant anymore? If so, these need to be removed these implementations from the library as well right?
> 
> You're right. I thought I had cleared all the `init` APIs. Will check again and will remove `clear_session` as well.

Also, need to update the `docs` as well.

File: crates/common/tedge_config/tests/test_tedge_config.rs [hunk 2]
```diff
@@ -753,16 +753,14 @@ fn test_parse_invalid_toml_file() -> Result<(), TEdgeConfigError> {
...
         Err(error @ TEdgeConfigError::Figment(_)) => {
-            assert_eq!(
-                error.to_string(),
-                format!(
-                    "unexpected character found: `<` at line 2 column 9 in {toml_path} TOML file"
-                )
+            assert!(
+                error.to_string().contains("invalid key"),
+                "error: {} does not contain the expected text: invalid key",
+                error
```
Comment (didier-wenzek) [inline crates/common/tedge_config/tests/test_tedge_config.rs:763]: Why the error returned by tedge_config has been changed?

Looking at the tested file content (`<abcde>`), the previous error message makes sense but not the new one.

File: docs/src/howto-guides/026_how_to_install_thin_edge_manually.md [hunk 3]
```diff
@@ -106,10 +106,8 @@ Next, create the files and directories required by thin-edge.io and restart mosq
 
 ```shell
 sudo rc-service mosquitto stop
-sudo tedge --init
+sudo tedge init
```
Comment (didier-wenzek) [inline docs/src/howto-guides/026_how_to_install_thin_edge_manually.md:109]: I would make the tedge user explicit here to give the user a hint that he can do differently.

```suggestion
sudo tedge init --user tedge --group tedge
```

**Review by didier-wenzek — COMMENTED** [review]
Nice move forward!

I would be really happy to approve. I have a doubt though regarding the test that has been changed in tedge_config. Why this fix?

File: configuration/debian/c8y-firmware-plugin/postinst [hunk 4]
```diff
@@ -1,8 +1,3 @@
 #!/bin/sh
 
-set -e
```
Comment (reubenmiller) [inline configuration/debian/c8y-firmware-plugin/postinst:3]: We shouldn't remove `set -e` here as the script should stop on unexpected errors (e.g. during the `#DEBHELPER#` section that is injected by cargo-deb)

**Review by rina23q — COMMENTED** [review]
Looks nice!

File: crates/common/tedge_config/tests/test_tedge_config.rs (hunk 2 above)
Comment (albinsuresh) [reply crates/common/tedge_config/tests/test_tedge_config.rs:763]: The error message changed in the latest version of the `figment` crate. So, I was forced to update this test. I personally feel that asserting exact error messages is a bit fragile due to issues like this. But didn't want to remove the whole assertion either, so decided to assert on just the significant part of the error message.

File: crates/common/tedge_utils/src/file.rs [hunk 5]
```diff
@@ -56,6 +56,20 @@ pub enum FileError {
...
+    dir: P,
+    user: Option<String>,
+    group: Option<String>,
+    mode: Option<u32>,
+) -> Result<(), FileError> {
+    let perm_entry = PermissionEntry::new(user, group, mode);
+    perm_entry.create_directory(dir.as_ref())
+}
+
+pub fn create_directory_with_defaults<P: AsRef<Path>>(dir: P) -> Result<(), FileError> {
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/tedge_utils/src/file.rs:69]: The user running the command. Any suggestions for a better name?

File: crates/common/tedge_utils/src/file.rs [hunk 6]
```diff
@@ -56,6 +56,20 @@ pub enum FileError {
     NoWriteAccess { path: PathBuf },
 }
 
+pub fn create_directory<P: AsRef<Path>>(
+    dir: P,
+    user: Option<String>,
+    group: Option<String>,
+    mode: Option<u32>,
+) -> Result<(), FileError> {
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/tedge_utils/src/file.rs:64]: With `create_directory_with_defaults`, I was just trying to provide an even easier API where you don't even have to type `PermissionEntry::default()` all the time. But yeah, I could live with `create_directory` as well.

File: docs/src/references/thin-edge-config-files.md [hunk 7]
```diff
@@ -1,19 +1,19 @@
 # Thin-edge config files
 
-Thin-edge.io requires config files for its operation. The `tedge --init` option is used to create
+Thin-edge.io requires config files for its operation. The `tedge init` command is used to create
 the base directory and other directories inside the base directory with appropriate user and permissions.
-The `tedge-mapper --init c8y/az` and `tedge-agent --init` will create the
-directories/files inside the base directory that are required for their operation.
+The `tedge-mapper c8y/az` and `tedge-agent` will create the config
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply docs/src/references/thin-edge-config-files.md:5]: It doesn't. I think it was just a generalised doc statement.

File: crates/extensions/c8y_firmware_manager/src/tests.rs [hunk 8]
```diff
@@ -317,105 +317,6 @@ async fn create_download_request_with_c8y_auth() -> Result<(), DynError> {
     Ok(())
 }
 
-#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/c8y_firmware_manager/src/tests.rs:320]: Deleting the file still won't result in any error (as written in this test) as with these changes, the directories will be recreated if it doesn't exist already. That's why I removed the whole test itself. May be I should add another test that deletes all these directories and checks if everything is fine on restart of the mapper.

File: configuration/debian/c8y-firmware-plugin/postinst (hunk 4 above)
Comment (albinsuresh) [reply configuration/debian/c8y-firmware-plugin/postinst:3]: If all that this script is doing is running `#DEBHELPER#`, is it even needed? Or a `postint` script will be  auto-generated in the absence of a custom one?

**Review by didier-wenzek — APPROVED** [review]
Approved. Thank you for these improvements.

File: configuration/debian/c8y-firmware-plugin/postinst (hunk 4 above)
Comment (reubenmiller) [reply configuration/debian/c8y-firmware-plugin/postinst:3]: Good question. From memory it is auto generated if the postinst file does not exist, so maybe that is the better option. You can verify the behaviour by just inspecting the deb file afterwards

File: crates/common/tedge_utils/src/file.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/tedge_utils/src/file.rs:64]: I updated the `create_directory` signature but decided to keep `create_directory_with_defaults` as it still helps with minimal typing.

**Comment by albinsuresh** [comment]
> Got it, but when all the services are on the same `node`, If the user gives different `group` names won't this affect creating the specific directories/files?

@PradeepKiruvale Only the `tedge init` command accepts the `user` and `group` parameters so that the top level directories get created with that owner. The other daemons don't have similar `--user` and `--group` options. So, there isn't the possibility of them init-ing different daemons differently. But, even with that single `tedge init`, the expectation is that all the daemons will be run by that same user (in the service definitions etc). If people mess up there, they'll be greeted with permission issues, which is expected.

**Pull request merged**